# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/html/nsHTMLDocument.cpp
# Commit: ddf4012a7652
# Full Hash: ddf4012a7652e36d144fc43bc99335276deeafbe
# Author: Christoph Kerschbaumer <ckerschb@christophkerschbaumer.com>
# Date: 2019-05-22 03:40:58
# Regressor Bug: 965637
# File Overlap Count: 1
# Description:
#   Bug 965637: Move CSP from Principal into Client, part 1: backend changes. r=mccr8
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D27654
# ==============================================================================

diff -r 6c2047bc2f3e -r ddf4012a7652 dom/html/nsHTMLDocument.cpp
--- a/dom/html/nsHTMLDocument.cpp	Tue May 21 22:51:54 2019 +0000
+++ b/dom/html/nsHTMLDocument.cpp	Tue May 21 23:14:27 2019 +0000
@@ -18,6 +18,7 @@
 #include "nsPrintfCString.h"
 #include "nsReadableUtils.h"
 #include "nsUnicharUtils.h"
+#include "nsIContentSecurityPolicy.h"
 #include "nsGlobalWindowInner.h"
 #include "nsIDocumentLoader.h"
 #include "nsIHTMLContentSink.h"
@@ -102,6 +103,7 @@
 #include "nsIImageDocument.h"
 #include "mozilla/dom/HTMLBodyElement.h"
 #include "mozilla/dom/HTMLDocumentBinding.h"
+#include "mozilla/dom/nsCSPContext.h"
 #include "mozilla/dom/Selection.h"
 #include "mozilla/dom/ShadowIncludingTreeIterator.h"
 #include "nsCharsetSource.h"
@@ -1043,6 +1045,17 @@
     }
   }
 
+  // document.open() inherits the CSP from the opening document.
+  // Please create an actual copy of the CSP (do not share the same
+  // reference) otherwise appending a new policy within the opened
+  // document will be incorrectly propagated to the opening doc.
+  nsCOMPtr<nsIContentSecurityPolicy> csp = callerDoc->GetCsp();
+  if (csp) {
+    RefPtr<nsCSPContext> cspToInherit = new nsCSPContext();
+    cspToInherit->InitFromOther(static_cast<nsCSPContext*>(csp.get()));
+    mCSP = cspToInherit;
+  }
+
   // At this point we know this is a valid-enough document.open() call
   // and not a no-op.  Increment our use counter.
   SetDocumentAndPageUseCounter(eUseCounter_custom_DocumentOpen);